src/utils/loop_utils.hpp

```diff
@@ -0,0 +1,137 @@
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef UTILS_LOOP_UTILS_HPP_
+#define UTILS_LOOP_UTILS_HPP_
+
+#include <memory>      // smart pointers
+#include <type_traits> // std::enable_if
+#include <utility>     // std::forward
+
+#include "bvals/cc/bnd_info.hpp" // TODO(JMM): Remove me when possible
+#include "mesh/domain.hpp"       // TODO(JMM): Remove me when possible
+
+namespace parthenon {
+
+// forward declarations
+class MeshBlock;
+template <typename T>
+class MeshBlockData;
+template <typename T>
+class MeshData;
+template <typename T>
+class CellVariable;
+class NeighborBlock;
+
+namespace loops {
+namespace shorthands {
+using sp_mb_t = std::shared_ptr<MeshBlock>;
+using sp_mbd_t = std::shared_ptr<MeshBlockData<Real>>;
+using sp_cv_t = std::shared_ptr<CellVariable<Real>>;
+using nb_t = NeighborBlock;
+} // namespace shorthands
+
+enum class LoopControl { cont, break_out };
+
+// Methods for wrapping a function that may or may not return a LoopControl
+// object. The first is enabled if the function returns a LoopControl and
+// just passes the returned object on. The second just calls the function,
+// ignores its return, and returns a LoopControl continue. These wrap the
+// function calls in the ForEachBoundary loop template to allow for breaking
+// out of the loop if desired
+template <class F, class... Args>
+inline auto func_caller(F func, Args &&...args) -> typename std::enable_if<
+    std::is_same<decltype(func(std::declval<Args>()...)), LoopControl>::value,
+    LoopControl>::type {
+  return func(std::forward<Args>(args)...);
+}
+
+template <class F, class... Args>
+inline auto func_caller(F func, Args &&...args) -> typename std::enable_if<
+    !std::is_same<decltype(func(std::declval<Args>()...)), LoopControl>::value,
+    LoopControl>::type {
+  func(std::forward<Args>(args)...);
+  return LoopControl::cont;
+}
+
+// Loop over boundaries (or shared geometric elements) for blocks contained
+// in MeshData, calling the passed function func for every boundary. Unifies
+// boundary looping that occurs in many places in the boundary communication
+// routines and allows for easy selection of a subset of the boundaries based
+// on the template parameter BoundaryType. [Really, this probably does not
+// need to be a template parameter, it could just be a function argument]
+template <BoundaryType bound = BoundaryType::any, class F>
+inline void ForEachBoundary(std::shared_ptr<MeshData<Real>> &md, F func) {
+  for (int block = 0; block < md->NumBlocks(); ++block) {
+    auto &rc = md->GetBlockData(block);
+    auto pmb = rc->GetBlockPointer();
+    for (auto &v : rc->GetCellVariableVector()) {
+      if (v->IsSet(Metadata::FillGhost)) {
+        for (int n = 0; n < pmb->pbval->nneighbor; ++n) {
```

Comment:
This is changed in #797, but I am guessing this PR will go in first and we can merge this PR into #797.